src/framework/coercion/core.clj

```diff
@@ -42,16 +43,30 @@
   on request error: responds {:status 400, :body \"Request coercion failed\"}
   on response error: responds {:status 400, :body \"Response validation failed\"}"
   {:enter (fn [state]
-            (let [cc (-> (get-in state [:request-data :match])
-                         coercion/coerce!)]
-              (xiana/ok (update-in state [:request :params] merge cc))))
-   :error (fn [state]
-            (xiana/error (assoc state :response {:status 400
-                                                 :body   "Request coercion failed"})))
-   :leave (fn [{{:keys [:status :body]} :response
-                :as                     state}]
-            (let [schema (get-in state [:request-data :match :data :responses status :body])]
-              (cond (and schema body (m/validate schema body)) (xiana/ok state)
-                    (and schema body) (xiana/error (assoc state :response {:status 400
-                                                                           :body   "Response validation failed"}))
-                    :else (xiana/ok state))))})
+            (try (let [cc (-> (get-in state [:request-data :match])
+                              coercion/coerce!)]
+                   (xiana/ok (update-in state [:request :params] merge cc)))
+                 (catch Exception ex ; TODO: be more specific
+                   (xiana/error (assoc state :response {:status 400
+                                                        :body {:errors
+                                                               (mapv (fn [e]
+                                                                       (format  "%s should satisfy %s"
+                                                                                (:in e)
+                                                                                (m/form (:schema e))))
+                                                                     (:errors (ex-data ex)))}})))))
+   :leave (fn [{{:keys [:status :body]}  :response
+                {method :request-method} :request
+                :as                      state}]
+            (let [schema (or (get-in state [:request-data :match :data method :responses status :body])
+                             (get-in state [:request-data :match :data :responses status :body]))
+                  valid? (some-> schema (m/validate body))]
+              (if (false? valid?)
+                (let [explain (m/explain schema body)
+                      humanized (me/humanize explain)]
```

Comment:
if response is not valid then this is an internal failure, we should log this error and only respond with 500.